### PR TITLE
refactor: Rework G4 surface mapping to make it more robust

### PR DESCRIPTION
**Pull request by benjaminhuth**
Should now work with ITk annulus bounds and be more robust in general. Also adds some dedicated debugging facilities to ease debugging in case of missing surfaces.

### CODE REVIEW

File: Examples/Algorithms/Geant4/src/Geant4Simulation.cpp [hunk 1]
```diff
@@ -282,11 +282,18 @@ ActsExamples::Geant4Simulation::Geant4Simulation(const Config& cfg,
...
+        sState, Acts::GeometryContext{}, false, false);
+    if (!allSurfacesMapped) {
+      ACTS_WARNING(
+          "Not all sensitive surfaces have been mapped to Geant4 volumes!");
+    }
 
     sensitiveSteppingActionAccess->assignSurfaceMapping(
         sState.g4VolumeToSurfaces);
+
+    throw "stop here";
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/Geant4Simulation.cpp:296]: 🛑

File: Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp [hunk 2]
```diff
@@ -9,15 +9,120 @@
...
 
 #include <G4LogicalVolume.hh>
 #include <G4Material.hh>
+#include <G4Polyhedron.hh>
 #include <G4VPhysicalVolume.hh>
+#include <G4VSolid.hh>
+#include <boost/container/flat_set.hpp>
+#include <boost/geometry.hpp>
+
+namespace boost::geometry::traits {
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp:31]: potentially one or two lines of comment would be nice to get an idea what these traits are about

File: Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp [hunk 3]
```diff
@@ -9,15 +9,120 @@
...
+void writeG4Polyhedron(Acts::ObjVisualization3D<T>& visualizer,
+                       const G4Polyhedron& polyhedron,
+                       Acts::Transform3 trafo = Acts::Transform3::Identity(),
+                       Acts::ColorRGB color = {0, 0, 0}) {
+  constexpr double convertLength = CLHEP::mm / Acts::UnitConstants::mm;
+
+  for (int i = 1; i <= polyhedron.GetNoFacets(); ++i) {
+    G4Point3D points[100];
+    int nPoints = 0;
+    polyhedron.GetFacet(i, nPoints, points);
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp:73]: should we check if 100 where enough? should 100 be a constant above?

File: Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp [hunk 4]
```diff
@@ -28,35 +133,47 @@ void ActsExamples::SensitiveSurfaceMapper::remapSensitiveNames(
     G4VPhysicalVolume* g4PhysicalVolume,
     const Acts::Transform3& motherTransform) const {
   // Make sure the unit conversion is correct
-  constexpr double convertLength = CLHEP::mm / Acts::UnitConstants::mm;
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp:31]: I think the constant could either stay here or we remove the comment above

File: Examples/Python/src/Geant4Component.cpp [hunk 5]
```diff
@@ -109,53 +135,53 @@ PYBIND11_MODULE(ActsPythonBindingsGeant4, mod) {
                   cfg, getDefaultLogger("SensitiveSurfaceMapper", level));
             }));
 
+    py::class_<State>(sm, "State").def(py::init<>());
```
Comment (andiwand) [inline Examples/Python/src/Geant4Component.cpp:138]: why do we need to wire the state out?

File: Examples/Algorithms/Geant4/src/Geant4Simulation.cpp (hunk 1 above)
Comment (benjaminhuth) [reply Examples/Algorithms/Geant4/src/Geant4Simulation.cpp:296]: ah good catch :D

File: Examples/Python/src/Geant4Component.cpp (hunk 5 above)
Comment (benjaminhuth) [reply Examples/Python/src/Geant4Component.cpp:138]: Hmm its helps testing the surface mapping outside of the simulation, which was convenient for debugging

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/SensitiveSurfaceMapper.hpp [hunk 6]
```diff
@@ -118,6 +122,20 @@ class SensitiveSurfaceMapper {
...
+  /// @param state state object after a call to remapSensitiveNames
+  /// @param gctx the geometry context
+  /// @param writeMissingG4VolsAsObj write the Geant4 volumes that are
+  /// not mapped to an obj file
+  /// @param writeMissingSurfacesAsObj write the sensitive surfaces that
+  /// where not mapped to an obj file
+  /// @return Returns true only if all sensitive surfaces where mapped
+  bool checkMapping(const State& state, const Acts::GeometryContext& gctx,
+                    bool writeMissingG4VolsAsObj = false,
+                    bool writeMissingSurfacesAsObj = false) const;
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/include/ActsExamples/Geant4/SensitiveSurfaceMapper.hpp:137]: Just looking at the interface I would wonder where these files are written and why I don't have the control over that. I guess this is mostly a debugging utility so maybe we should hide the parameters from the outside and have just the flags and pathnames inside the function?

File: Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp [hunk 7]
```diff
@@ -9,15 +9,126 @@
...
+  static inline void set(Point& p, CoordinateType const& value) {
+    p[Index] = value;
+  }
+};
+
+}  // namespace boost::geometry::traits
+
+namespace {
+template <typename T>
+void writeG4Polyhedron(Acts::ObjVisualization3D<T>& visualizer,
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/src/SensitiveSurfaceMapper.cpp:66]: can this be un-templated by relying on the base class `IVisualization3D`?

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/SensitiveSurfaceMapper.hpp (hunk 6 above)
Comment (benjaminhuth) [reply Examples/Algorithms/Geant4/include/ActsExamples/Geant4/SensitiveSurfaceMapper.hpp:137]: Hmm so the check mapping function is now also called within the Simulation now, so I need some flags to controll the writing of the files...

**Review by andiwand — APPROVED** [review]

**Comment by asalzburger** [comment]
There's a `histcmp` failure ... @benjaminhuth do you know what this could be?

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [3b32e5d816413ec0c88ebbe8734d9f48b4d223d6]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879260)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879259)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879258)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879257)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879256)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879255)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879254)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879253)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879252)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879251)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879250)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879249)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879248)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879247)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879246)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879245)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879244)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879243)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879242)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879241)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879240)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879239)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879238)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879237)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879236)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879235)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879234)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879233)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879232)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879231)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879230)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879229)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42879228)||

**Pull request merged**